great_expectations/data_context/data_context/abstract_data_context.py

```diff
@@ -1575,11 +1649,83 @@ def update_checkpoint(self, checkpoint: Checkpoint) -> Checkpoint:
         """
         return self.checkpoint_store.update_checkpoint(checkpoint)
 
+    @overload
+    def add_or_update_checkpoint(
+        self,
+        name: str = ...,
+        config_version: int | float | None = ...,
+        template_name: str | None = ...,
+        module_name: str | None = ...,
+        class_name: str | None = ...,
+        run_name_template: str | None = ...,
+        expectation_suite_name: str | None = ...,
+        batch_request: dict | None = ...,
+        action_list: list[dict] | None = ...,
+        evaluation_parameters: dict | None = ...,
+        runtime_configuration: dict | None = ...,
+        validations: list[dict] | None = ...,
+        profilers: list[dict] | None = ...,
+        # Next two fields are for LegacyCheckpoint configuration
+        validation_operator_name: str | None = ...,
+        batches: list[dict] | None = ...,
+        # the following four arguments are used by SimpleCheckpoint
+        site_names: str | list[str] | None = ...,
+        slack_webhook: str | None = ...,
+        notify_on: str | None = ...,
+        notify_with: str | list[str] | None = ...,
+        ge_cloud_id: str | None = ...,
+        expectation_suite_ge_cloud_id: str | None = ...,
+        default_validation_id: str | None = ...,
+        id: str | None = ...,
+        expectation_suite_id: str | None = ...,
+        checkpoint: None = ...,
+    ) -> Checkpoint:
+        """
+        Individual constructor arguments are provided.
+        `checkpoint` should not be provided.
+        """
+        ...
+
+    @overload
+    def add_or_update_checkpoint(
+        self,
+        name: None = ...,
+        config_version: None = ...,
+        template_name: None = ...,
+        module_name: None = ...,
+        class_name: None = ...,
+        run_name_template: None = ...,
+        expectation_suite_name: None = ...,
+        batch_request: None = ...,
+        action_list: None = ...,
+        evaluation_parameters: None = ...,
+        runtime_configuration: None = ...,
+        validations: None = ...,
+        profilers: None = ...,
+        validation_operator_name: None = ...,
+        batches: None = ...,
```

Comment:
I would say this is definitely out of scope for this PR, but at a later date we might want to replace all the parameters with an "options" or "parameters" type?
Like a `dataclass`, `NamedTuple` or even `TypeDict`.
```python
def add_or_update_checkpoint(
        self,
        options: UpdateCheckpointOptions | None = None,
        checkpoint: Checkpoint | None = None,
    ) -> Checkpoint:
```
https://refactoring.guru/smells/long-parameter-list
https://refactoring.guru/introduce-parameter-object

Reply:
Oh I'd love that - unfortunately, this is super legacy and an entrenched pattern so it'll be a breaking change. We could do it with a deprecation strategy though